File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-user-store/src/state/userStoreState.ts
import type { UserStorePageResult } from "../types/userStoreModels.js";

export interface UserStoreState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: UserStorePageResult<unknown>;
  readonly error?: string;
}

export function initialUserStoreState(): UserStoreState {
  return { status: "idle" };
}

export function reduceUserStoreState(
  state: UserStoreState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: UserStorePageResult<unknown> }
    | { type: "failed"; error: string },
): UserStoreState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}

File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_user_store/lib/src/state/user_store_state.dart
import '../models/user_store_models.dart';

enum UserStoreStatus { idle, loading, ready, error }

class UserStoreState {
  const UserStoreState({
    this.status = UserStoreStatus.idle,
    this.lastResult,
    this.error,
  });

  final UserStoreStatus status;
  final UserStorePageResult<Object?>? lastResult;
  final String? error;

  UserStoreState loading() => const UserStoreState(status: UserStoreStatus.loading);

  UserStoreState loaded(UserStorePageResult<Object?> result) =>
      UserStoreState(status: UserStoreStatus.ready, lastResult: result);

  UserStoreState failed(String message) =>
      UserStoreState(status: UserStoreStatus.error, error: message);
}
